### PR TITLE
Perf: Pairing on BN254 using direct Fp12 extension and non-native `Eval()`

### PR DESCRIPTION
**Pull request by yelhousni**
# Description

Similar to #1312 but for BN254. 
This PR refactor methods in `fields_bn254` and `sw_bn254` using a direct Fp12 extension and non-native `Eval()` introduced in #1299.

TODO:
- [x] Granger-Scott's cyclotomic squarings.
- [ ] ~~Karabina's cyclotomic squarings.~~
- [ ] ~~Torus-based arithmetic over direct Fp6 extension.~~

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- New tests corresponding to new methods.
- PairingCheck and FinalExpCheck tests against gnark-crypto.

# How has this been benchmarked?

Some circuits proved in a in a BN254-PLONK:

| circuit | old (scs) | new (scs) | diff (scs)
| ------ | --- | ---- | - |
| `e(a,b)` | 3,534,755 | 2,030,447 | 1,504,308 |
| `e(a,b) * e(c,d) == 1` |  3,879,428 |  2,239,682 | 1,639,746 |
| ECPAIR-2 precompile* | 4,407,402 | 2,425,399 | 1,982,003 |
| ECPAIR-4 precompile* | 7,815,376 | 4,181,633 | 3,633,743 |

*ECPAIR precompiles include G2 subgroup membership.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/algebra/emulated/sw_bn254/pairing.go b/std/algebra/emulated/sw_bn254/pairing.go
index fdfedafe..d2e95540 100644
--- a/std/algebra/emulated/sw_bn254/pairing.go
+++ b/std/algebra/emulated/sw_bn254/pairing.go
@@ -24,7 +24,6 @@ type Pairing struct {
 	curve  *sw_emulated.Curve[BaseField, ScalarField]
 	g2     *G2
 	bTwist *fields_bn254.E2
-	g2gen  *G2Affine
 }
 
 func NewGTEl(a bn254.GT) GTEl {
@@ -82,15 +81,6 @@ func NewPairing(api frontend.API) (*Pairing, error) {
 	}, nil
 }
 
-func (pr Pairing) generators() *G2Affine {
-	if pr.g2gen == nil {
-		_, _, _, g2gen := bn254.Generators()
-		cg2gen := NewG2AffineFixed(g2gen)
-		pr.g2gen = &cg2gen
-	}
-	return pr.g2gen
-}
-
 // Pair calculates the reduced pairing for a set of points
 // ∏ᵢ e(Pᵢ, Qᵢ).
 //

```

File: std/algebra/emulated/fields_bn254/hints.go [hunk 1]
```diff
@@ -186,34 +187,67 @@ func divE6By6Hint(nativeMod *big.Int, nativeInputs, nativeOutputs []*big.Int) er
 func inverseE12Hint(nativeMod *big.Int, nativeInputs, nativeOutputs []*big.Int) error {
 	return emulated.UnwrapHint(nativeInputs, nativeOutputs,
 		func(mod *big.Int, inputs, outputs []*big.Int) error {
+			var d [12]big.Int
```
Comment (ivokub) [inline std/algebra/emulated/fields_bn254/hints.go:190]: I would refactor mapping into separate function and then call here and in divE12Hint. Same for the reverse mapping.

**Review by ivokub — APPROVED** [review]
See the suggested edits for removing deadcode.

Looks good to me, I didn't go through the formulas, but as the tests are functioning, then I think everything is valid. See only comments about removed examples and tests.

One thing I would do is to create isomorphism maps between the tower and direct extension as explicit methods and use them, as imo it makes it a bit clearer to see what extension is used where.

And I think it would be better to wait for https://github.com/Consensys/gnark/pull/1340 to be merged to ensure that there are no failing tests for edge cases (which I understood you bypassed?)

File: std/algebra/emulated/sw_bn254/doc_test.go [hunk 2]
```diff
@@ -1,93 +0,0 @@
-package sw_bn254_test
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/emulated/sw_bn254/doc_test.go:1]: I removed a lot of things a the beginning to start from scratch and forgot to add back.

File: std/algebra/emulated/sw_bn254/pairing_test.go [hunk 3]
```diff
@@ -180,73 +128,6 @@ func TestPairTestSolve(t *testing.T) {
 	assert.NoError(err)
 }
 
-func TestPairFixedTestSolve(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/emulated/sw_bn254/pairing_test.go:183]: Will add back.

File: std/algebra/emulated/sw_bn254/pairing_test.go [hunk 4]
```diff
@@ -419,65 +300,6 @@ func TestIsOnG2Solve(t *testing.T) {
 	assert.NoError(err)
 }
 
-type IsMillerLoopAndFinalExpCircuit struct {
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/emulated/sw_bn254/pairing_test.go:422]: Will add back.

**Comment by yelhousni** [comment]
> See the suggested edits for removing deadcode.
> 
> Looks good to me, I didn't go through the formulas, but as the tests are functioning, then I think everything is valid. See only comments about removed examples and tests.
> 
> One thing I would do is to create isomorphism maps between the tower and direct extension as explicit methods and use them, as imo it makes it a bit clearer to see what extension is used where.
> 
> And I think it would be better to wait for #1340 to be merged to ensure that there are no failing tests for edge cases (which I understood you bypassed?)

That makes sense. I pushed few comments that address your review points. We can wait for https://github.com/Consensys/gnark/pull/1340. 
I'm still thing whether to implement Karabina and Torus-based direct-Fp6 as it would be only useful for computing the exact value of the pairing, but we only use `PairingCheck` in our protocols. However it might come in handy in the future for Ethereum SSLE ([ethresearch post](https://ethresear.ch/t/the-return-of-torus-based-cryptography-whisk-and-curdleproof-in-the-target-group/16678/4?u=yelhousni)).

**Comment by ivokub** [comment]
> That makes sense. I pushed a few comments that address your review points. We can wait for #1340. I'm still thinking whether to implement Karabina and Torus-based direct-Fp6 as it would be only useful for computing the exact value of the pairing, but we only use `PairingCheck` in our protocols. However torus stuff might come in handy in the future for Ethereum SSLE ([ethresearch post](https://ethresear.ch/t/the-return-of-torus-based-cryptography-whisk-and-curdleproof-in-the-target-group/16678/4?u=yelhousni)).

Yup, I agree that with different extensions computing pairing makes less sense and we'd mostly be interested in pairing check (or multi-pairing check). I guess we can keep it currently as is, but have documentation update that it may not be efficient.

**Comment by yelhousni** [comment]
@ivokub PR is ready now!

**Comment by ivokub** [comment]
Suggested edit:
```diff
diff --git a/std/algebra/emulated/sw_emulated/point.go b/std/algebra/emulated/sw_emulated/point.go
index 10b27de3..7fce40b4 100644
--- a/std/algebra/emulated/sw_emulated/point.go
+++ b/std/algebra/emulated/sw_emulated/point.go
@@ -214,13 +214,15 @@ func (c *Curve[B, S]) AssertIsOnCurve(p *AffinePoint[B]) {
 	selector := c.api.And(c.baseApi.IsZero(&p.X), c.baseApi.IsZero(&p.Y))
 	b := c.baseApi.Select(selector, c.baseApi.Zero(), &c.b)
 
-	left := c.baseApi.Mul(&p.Y, &p.Y)
-	right := c.baseApi.Eval([][]*emulated.Element[B]{{&p.X, &p.X, &p.X}, {b}}, []int{1, 1})
-	if c.addA {
-		ax := c.baseApi.Mul(&c.a, &p.X)
-		right = c.baseApi.Add(right, ax)
+	mone := c.baseApi.NewElement(-1)
+
+	var check *emulated.Element[B]
+	if !c.addA {
+		check = c.baseApi.Eval([][]*emulated.Element[B]{{&p.X, &p.X, &p.X}, {b}, {mone, &p.Y, &p.Y}}, []int{1, 1, 1})
+	} else {
+		check = c.baseApi.Eval([][]*emulated.Element[B]{{&p.X, &p.X, &p.X}, {&c.a, &p.X}, {b}, {mone, &p.Y, &p.Y}}, []int{1, 1, 1, 1})
 	}
-	c.baseApi.AssertIsEqual(left, right)
+	c.baseApi.AssertIsEqual(check, c.baseApi.Zero())
 }
 
 // AddUnified adds p and q and returns it. It doesn't modify p nor q.

```

**Review by ivokub — APPROVED** [review]
Thanks! Now looks very nice. I also found one optimization for AssertIsOnCurve where we can bring the left hand side to the right and then do only a single eval.
